src/main/kotlin/uk/gov/dluhc/printapi/service/PrintResponseProcessingService.kt

```diff
@@ -0,0 +1,92 @@
+import uk.gov.dluhc.printapi.database.entity.Status.NOT_DELIVERED
+import uk.gov.dluhc.printapi.database.entity.Status.PENDING_ASSIGNMENT_TO_BATCH
+import uk.gov.dluhc.printapi.database.entity.Status.PRINT_PROVIDER_DISPATCH_FAILED
+import uk.gov.dluhc.printapi.database.entity.Status.PRINT_PROVIDER_PRODUCTION_FAILED
+import uk.gov.dluhc.printapi.database.entity.Status.PRINT_PROVIDER_VALIDATION_FAILED
+import uk.gov.dluhc.printapi.database.entity.Status.RECEIVED_BY_PRINT_PROVIDER
+import uk.gov.dluhc.printapi.database.entity.Status.SENT_TO_PRINT_PROVIDER
+import uk.gov.dluhc.printapi.database.entity.Status.VALIDATED_BY_PRINT_PROVIDER
+import uk.gov.dluhc.printapi.database.repository.PrintDetailsRepository
+import uk.gov.dluhc.printapi.messaging.models.ProcessPrintResponseMessage
+import uk.gov.dluhc.printapi.printprovider.models.BatchResponse
+import uk.gov.dluhc.printapi.printprovider.models.BatchResponse.Status.SUCCESS
+import java.time.Clock
+import java.time.OffsetDateTime
+import uk.gov.dluhc.printapi.messaging.models.ProcessPrintResponseMessage.Status as ResponseStatus
+import uk.gov.dluhc.printapi.messaging.models.ProcessPrintResponseMessage.StatusStep as ResponseStatusStep
+
+class PrintResponseProcessingService(
+    private val printDetailsRepository: PrintDetailsRepository,
+    private val idFactory: IdFactory,
+    private val clock: Clock
+) {
+    /**
+     * Updates print details for the batches based on the batch responses' statuses.
+     * If a batch is successful, a new printRequestStatus will be added for each corresponding print details with status
+     * being RECEIVED_BY_PRINT_PROVIDER.
+     *
+     * If a batch is not successful, the status will be set to SENT_TO_PRINT_PROVIDER and a new requestId will be set.
+     * The reason for the new requestId is, the print provider does not support duplicate print requests with the same requestId.
+     *
+     * dateCreated will be the timestamp the new printRequestStatus is created and eventDateTime will be the timestamp from
+     * the print provider's batch response.
+     */
+    fun processBatchResponses(batchResponses: List<BatchResponse>) {
+        batchResponses.forEach { batchResponse ->
+            with(batchResponse) {
+                val newStatus =
+                    if (status == SUCCESS) RECEIVED_BY_PRINT_PROVIDER else PENDING_ASSIGNMENT_TO_BATCH
+
+                val printDetails =
+                    printDetailsRepository.getAllByStatusAndBatchId(SENT_TO_PRINT_PROVIDER, batchId)
+
+                printDetails.forEach {
+                    it.addStatus(
+                        status = newStatus,
+                        dateCreated = OffsetDateTime.now(clock),
+                        eventDateTime = timestamp,
+                        message = null
+                    )
+
+                    if (status != SUCCESS) {
+                        it.requestId = idFactory.requestId()
+                    }
```

Comment:
I know it's not in the story, but based on a problem we had yesterday can we also null out the batchId in this if block please (so the PrintDetails instance has a status of PENDING_ASSIGNMENT_TO_BATCH and a null batchId)
```
                    if (status != SUCCESS) {
                        it.requestId = idFactory.requestId()
                        it.batchId = null
                    }
```

Reply:
Ah, one other thing in the case of a `FAILED` status; we also need to map the `message` field from the `BatchResponse` into the new `PrintResponseStatus` object on line 52 above

Reply:
Thanks, @NathanRussellValtech ,
I will address the issues.